code/client/cl_http_windows.c

```diff
@@ -0,0 +1,162 @@
+published by the Free Software Foundation; either version 2 of the License,
+or (at your option) any later version.
+
+Quake III Arena source code is distributed in the hope that it will be
+useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with Quake III Arena source code; if not, write to the Free Software
+Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+===========================================================================
+*/
+
+#ifdef USE_HTTP
+
+#include "client.h"
+
+#include <windows.h>
+#include <wininet.h>
+
+static HINTERNET hInternet = NULL;
+static HINTERNET hUrl = NULL;
+
+static void DropIf(qboolean condition, const char *fmt, ...)
+{
+    char buffer[1024];
+
+    if (!condition)
+        return;
+
+    va_list argptr;
+    va_start(argptr, fmt);
+    Q_vsnprintf(buffer, sizeof(buffer), fmt, argptr);
+    va_end(argptr);
+
+    Com_Error(ERR_DROP, "Download Error: %s URL: %s", buffer, clc.downloadURL);
+}
+
+/*
+=================
+CL_HTTP_Init
+=================
+*/
+qboolean CL_HTTP_Init()
+{
+    OSVERSIONINFO osvi = {sizeof(OSVERSIONINFO)};
+    const char *windowsVersion = GetVersionEx(&osvi) ? va("Windows %lu.%lu (build %lu)",
+                                                          osvi.dwMajorVersion,
+                                                          osvi.dwMinorVersion,
+                                                          osvi.dwBuildNumber)
+                                                     : "Windows";
+
+    hInternet = InternetOpenA(
+        va("%s %s", Q3_VERSION, windowsVersion),
+        INTERNET_OPEN_TYPE_PRECONFIG, NULL, NULL, 0);
+
+    return hInternet != NULL;
+}
+
+/*
+=================
+CL_HTTP_Available
+=================
+*/
+qboolean CL_HTTP_Available()
+{
+    return hInternet != NULL;
+}
+
+/*
+=================
+CL_HTTP_Shutdown
+=================
+*/
+void CL_HTTP_Shutdown(void)
+{
+    if (hInternet)
+    {
+        InternetCloseHandle(hInternet);
+        hInternet = NULL;
+    }
+}
+
+/*
+=================
+CL_HTTP_BeginDownload
+=================
+*/
+void CL_HTTP_BeginDownload(const char *remoteURL)
+{
+    DWORD httpCode = 0;
+    DWORD contentLength = 0;
+    DWORD len = sizeof(httpCode);
+    DWORD zero = 0;
+    BOOL success;
+
+    hUrl = InternetOpenUrlA(hInternet, remoteURL, NULL, 0,
+                            INTERNET_FLAG_HYPERLINK |
+                                INTERNET_FLAG_NO_CACHE_WRITE |
+                                INTERNET_FLAG_NO_COOKIES |
+                                INTERNET_FLAG_NO_UI |
+                                INTERNET_FLAG_RESYNCHRONIZE |
+                                INTERNET_FLAG_RELOAD |
+                                INTERNET_FLAG_SECURE,
```

Comment:
I think the INTERNET_FLAG_SECURE flag should be removed from InternetOpenUrlA(). The [documentation says](https://learn.microsoft.com/en-us/windows/win32/wininet/api-flags) it's redundant for https URLs and only applies to http which sounds like it forces http to use https.

Reply:
TBH I just (mostly) blindly copied the flags used in the autoupdater. Good catch.